californium-core/src/main/java/org/eclipse/californium/core/network/serialization/DataParser.java

```diff
@@ -1,222 +1,114 @@
-
-	/**
-	 * Tries to read a token, options and payload from the binary representation of the CoAP message
-	 * this parser has been created for.
-	 * 
-	 * @param message the CoAP message object to set the read values on.
-	 * @throws MessageFormatException if the underlying message contains a message format error.
-	 */
-	private void parseMessage(final Message message) {
-		assertCorrectVersion();
-		message.setType(Type.valueOf(type));
-		message.setMID(mid);
-
-		if (tokenlength > 0) {
-			message.setToken(reader.readBytes(tokenlength));
-		} else {
-			message.setToken(new byte[0]);
-		}
-
-		int currentOptionNumber = 0;
-		byte nextByte = 0;
-
-		// TODO detect malformed options
-		while (reader.bytesAvailable()) {
-			nextByte = reader.readNextByte();
-			if (nextByte != PAYLOAD_MARKER) {
-				// the first 4 bits of the byte represent the option delta
-				int optionDeltaNibble = (0xF0 & nextByte) >> 4;
-				currentOptionNumber = calculateNextOptionNumber(currentOptionNumber, optionDeltaNibble);
-
-				// the second 4 bits represent the option length
-				int optionLengthNibble = 0x0F & nextByte;
-				int optionLength = determineValueFromNibble(optionLengthNibble);
-
-				// read option
-				Option option = new Option(currentOptionNumber);
-				option.setValue(reader.readBytes(optionLength));
-
-				// add option to message
-				message.getOptions().addOption(option);
-			} else break;
-		}
-
-		if (nextByte == PAYLOAD_MARKER) {
-			// the presence of a marker followed by a zero-length payload must be processed as a message format error
-			if (!reader.bytesAvailable()) {
-				throw new MessageFormatException("Found payload marker (0xFF) but message contains no payload");
-			} else {
-				// get payload
-				message.setPayload(reader.readBytesLeft());
-			}
-		} else {
-			message.setPayload(new byte[0]); // or null?
-		}
-	}
-
-	/**
-	 * Calculates the next option number based on the current option number and the option delta as specified in
-	 * RFC 7252, Section 3.1
-	 * 
-	 * @param delta
-	 *            the 4-bit option delta value.
-	 * @return the next option number.
-	 * @throws MessageFormatException if the option number cannot be determined due to a message format error.
-	 */
-	private int calculateNextOptionNumber(final int currentOptionNumber, final int delta) {
-		return currentOptionNumber + determineValueFromNibble(delta);
-	}
-
-	private int determineValueFromNibble(final int delta) {
-		if (delta <= 12) {
-			return delta;
-		} else if (delta == 13) {
-			return reader.read(8) + 13;
-		} else if (delta == 14) {
-			return reader.read(16) + 269;
-		} else {
-			throw new MessageFormatException("Message contains illegal option delta/length: " + delta);
-		}
-	}
+public abstract class DataParser {
+
+    /** Generate RST for a specified input. */
+    public abstract EmptyMessage generateRst(RawData rawData);
+
+    /** Parses message based on the raw data. */
+    public Request parseRequest(RawData rawData, int code) {
```

Comment:
There's a bit of circular dependency between parsing the messages and knowing what type it is. Agreed, done this way is a bit counterintuitive.

I think the simpler way would be to not try and have common parse(Request|Response) methods on DataParser, and instead having (Udp|TCP)Parsers implement those directly, since they know how to extract code out of the message. A tiny bit of code duplication, but it's easier to understand.